src/cipher/mod.rs

```diff
@@ -0,0 +1,359 @@
+///
+///     // Successful communication
+///     assert_eq!(decrypted_message, message);
+/// }
+///
+/// fn sender(
+///     sender_secret: &JubJubScalar,
+///     receiver_public: &ExtendedPoint,
+///     message: &[BlsScalar],
+/// ) -> (PoseidonCipher, BlsScalar) {
+///     // Use the Diffie-Hellman protocol to generate a shared secret
+///     let shared_secret = dhke(sender_secret, receiver_public);
+///
+///     // Generate a random nonce that will be public
+///     let nonce = BlsScalar::random(&mut rand::thread_rng());
+///
+///     // Encrypt the message
+///     let cipher = PoseidonCipher::encrypt(&message, &shared_secret, &nonce);
+///
+///     (cipher, nonce)
+/// }
+///
+/// fn receiver(
+///     receiver_secret: &JubJubScalar,
+///     sender_public: &ExtendedPoint,
+///     cipher: &PoseidonCipher,
+///     nonce: &BlsScalar,
+/// ) -> [BlsScalar; MESSAGE_CAPACITY] {
+///     // Use the Diffie-Hellman protocol to generate a shared secret
+///     let shared_secret = dhke(receiver_secret, sender_public);
+///
+///     // Decrypt the message
+///     cipher
+///         .decrypt(&shared_secret, &nonce)
+///         .expect("Failed to decrypt!")
+/// }
+/// ```
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Ord, PartialOrd, Default)]
+pub struct PoseidonCipher {
+    cipher: [BlsScalar; CIPHER_SIZE],
+}
+
+impl PoseidonCipher {
+    /// [`PoseidonCipher`] constructor
+    pub fn new(cipher: [BlsScalar; CIPHER_SIZE]) -> Self {
+        Self { cipher }
+    }
+
+    /// Maximum number of scalars allowed per message
+    pub fn capacity() -> usize {
+        MESSAGE_CAPACITY
+    }
+
+    /// Bytes consumed on serialization of the poseidon cipher
+    pub fn serialized_size() -> usize {
+        ENCRYPTED_DATA_SIZE
+    }
+
+    /// Encrypt a slice of scalars into an internal cipher representation
+    ///
+    /// The message size will be truncated to [`MESSAGE_CAPACITY`] bits
+    pub fn encrypt(
+        message: &[BlsScalar],
+        secret: &AffinePoint,
+        nonce: &BlsScalar,
+    ) -> Self {
+        let zero = BlsScalar::zero();
+        let mut strategy = ScalarStrategy::new();
+
+        let mut cipher = [zero; CIPHER_SIZE];
+        let mut state = PoseidonCipher::initial_state(secret, *nonce);
+
+        strategy.perm(&mut state);
+
+        (0..MESSAGE_CAPACITY).for_each(|i| {
+            state[i + 1] += if i < message.len() {
+                message[i]
+            } else {
+                BlsScalar::zero()
+            };
+
+            cipher[i] = state[i + 1];
+        });
+
+        strategy.perm(&mut state);
+        cipher[MESSAGE_CAPACITY] = state[1];
+
+        PoseidonCipher::new(cipher)
+    }
+
+    /// Perform the decrypt of a previously encrypted message.
+    ///
+    /// Will return `None` if the decryption fails.
+    pub fn decrypt(
+        &self,
+        secret: &AffinePoint,
+        nonce: &BlsScalar,
+    ) -> Option<[BlsScalar; MESSAGE_CAPACITY]> {
```

Comment:
Why an `Option` instead of an `Error`??

Reply:
Done. Was just for the sake of not creating an error struct but having it is the best way.